src/wp-includes/class-wp-block-metadata-registry.php

```diff
@@ -0,0 +1,227 @@
+	 * directory name itself for directories.
+	 *
+	 * Example manifest file structure:
+	 * ```
+	 * return array(
+	 *     'example-block' => array(
+	 *         'title' => 'Example Block',
+	 *         'category' => 'widgets',
+	 *         'icon' => 'smiley',
+	 *         // ... other block metadata
+	 *     ),
+	 *     'another-block' => array(
+	 *         'title' => 'Another Block',
+	 *         'category' => 'formatting',
+	 *         'icon' => 'star-filled',
+	 *         // ... other block metadata
+	 *     ),
+	 *     // ... more block metadata entries
+	 * );
+	 * ```
+	 *
+	 * For more information on how the block identifiers are determined, see the
+	 * `default_identifier_callback()` method.
+	 *
+	 * @since 6.X.0
+	 *
+	 * @param string $path     The absolute base path for the collection ( e.g., WP_PLUGIN_DIR . '/my-plugin/blocks/' ).
+	 * @param string $manifest The absolute path to the manifest file containing the metadata collection.
+	 * @return bool            True if the collection was registered successfully, false otherwise.
+	 */
+	public static function register_collection( $path, $manifest ) {
+		$path = wp_normalize_path( rtrim( $path, '/' ) );
+
+		// Check if the path is valid:
+		if ( str_starts_with( $path, wp_normalize_path( ABSPATH . WPINC ) ) ) {
+			// Core path is valid.
+		} elseif ( str_starts_with( $path, wp_normalize_path( WP_PLUGIN_DIR ) ) ) {
+			// For plugins, ensure the path is within a specific plugin directory and not the base plugin directory.
+			$plugin_dir    = wp_normalize_path( WP_PLUGIN_DIR );
+			$relative_path = substr( $path, strlen( $plugin_dir ) + 1 );
+			$plugin_name   = strtok( $relative_path, '/' );
+
+			if ( empty( $plugin_name ) || $plugin_name === $relative_path ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Block metadata collections can only be registered for a specific plugin. The provided path must be within a plugin directory.' ),
+					'6.X.0'
+				);
+				return false;
+			}
+		} else {
+			_doing_it_wrong(
+				__METHOD__,
+				__( 'Block metadata collections can only be registered for a specific plugin. The provided path is neither a core path nor a valid plugin path.' ),
+				'6.X.0'
+			);
+			return false;
+		}
+
+		if ( ! file_exists( $manifest ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				__( 'The specified manifest file does not exist.' ),
+				'6.X.0'
+			);
+			return false;
+		}
+
+		self::$collections[ $path ] = array(
+			'manifest' => $manifest,
+			'metadata' => null,
+		);
+
+		return true;
+	}
+
+	/**
+	 * Retrieves block metadata for a given block within a specific collection.
+	 *
+	 * This method uses the registered collections to efficiently lookup
+	 * block metadata without reading individual `block.json` files.
+	 *
+	 * @since 6.X.0
+	 *
+	 * @param string $file_or_folder The path to the file or folder containing the block.
+	 * @return array|null            The block metadata for the block, or null if not found.
```

Comment:
See above. Please also check the rest of the PR.
```suggestion
	 * @return array|null The block metadata for the block, or null if not found.
```